### ISSUE TITLE
SFTTrainer usage

### ISSUE THREAD

**Issue by Humauaca** [huggingface/trl#2390]
### Method description

I want to fine-tune meta-llama/Llama-3.2-1B-Instruct with SFTTrainer, but I don't know how to process the dataset (custom dataset). If I'm not wrong, the inputs should be the sentence minus the last token, and the labels should be the sentence minus the first token. Is this automatically handled by the SFTTrainer, or do I have to do it manually in the dataset?
Thanks.

### Open source status

- [X] The method implementation is available
- [X] The model weights are available
- [ ] The training datasets are available

**Comment by qgallouedec** [comment]
You don't need to process the data. The trainer does it for you. The `SFTTrainer` expects a dataset with a column named `"text"` (or `"messages"` for conversational data). Use [trl-internal-testing/zen](https://huggingface.co/datasets/trl-internal-testing/zen) as an example.


```
from trl import SFTConfig, SFTTrainer
from datasets import load_dataset

dataset = load_dataset("trl-internal-testing/zen", "conversational_language_modeling", split="train")

training_args = SFTConfig(output_dir="Llama-3.2-1B-Instruct-SFT")
trainer = SFTTrainer(
    args=training_args,
    model="meta-llama/Llama-3.2-1B-Instruct",
    train_dataset=dataset,
)
trainer.train()
```

**Comment by qgallouedec** [comment]
Yes

**Comment by Alex-Mathai-98** [comment]
Hi @qgallouedec - just a simple follow-up question. In the conversation format, does the SFT Trainer mask out the loss for the instructions? Or does it compute the loss for both the instructions and the responses?

**Comment by qgallouedec** [comment]
Not by default, you need to use `DataCollatorForCompletionOnly`

**Comment by Alex-Mathai-98** [comment]
I see, @qgallouedec. Thank you so much for your response. I just want to make sure I understand your response perfectly. 

I was talking about https://huggingface.co/docs/trl/sft_trainer#dataset-format-support - so even if we format our data in this **conversations** format, HF will tokenize the entire conversation as a sequence of tokens and then perform **simple language modeling** on the entire text.

The link https://huggingface.co/docs/trl/sft_trainer#train-on-completions-only you mentioned in your response seems to be used for a use case that is **NOT** conversations right?

In a multi-turn conversation, there will be **multiple** answers - unlike the example mentioned in https://huggingface.co/docs/trl/sft_trainer#train-on-completions-only . In this case, the masking needs to be done for all intermediate instruction text right?

I guess I will need to do this with a custom data collator ?

**Comment by hiteshsom** [comment]
Hi

Does DataCollatorForCompletionOnlyLM

required template of instruction and response 
Or
name of the keys containing instruction and response in dataset ?

What I thought was that dataset will have one column ‘text’

Lets say one sample will look like:

“<|im_start|>system\nYou are a helpful assistant. <|im_end|><|im_start|>user\nWhat is 2+2? <|im_end><|im_start|>assistant\nThe answer is 4.<|im_end|>”

Now when using DataCollatorForCompletionOnlyLM

I can set
instruction_template = “<|im_start|>system\n”
response_template = “<|im_start|>assistant\n”

and the Data Collator will auto consider to mask
You are a helpful assistant …. What is 2 + 2 ?

during loss and calculate loss only while predicting 

The answer is 4

But after reading the code:

https://github.com/huggingface/trl/blob/main/trl/trainer/utils.py

line 95 to line 109

```python

self.instruction_template = instruction_template
        if isinstance(instruction_template, str):
            # The user provides a string, must tokenize
            self.instruction_token_ids = self.tokenizer.encode(self.instruction_template, add_special_tokens=False)
        else:
            # The user already provides the token ids
            self.instruction_token_ids = instruction_template

        self.response_template = response_template
        if isinstance(response_template, str):
            # The user provides a string, must tokenize
            self.response_token_ids = self.tokenizer.encode(self.response_template, add_special_tokens=False)
        else:
            # The user already provides the token ids
            self.response_token_ids = response_template


``` 

It seems that the instruction and response templates are passed as it is to the tokenizer, instead of searching for a template matching logic.

Am I wrong ?
